compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmBuiltinOptimizationLowering.kt

```diff
@@ -93,6 +95,66 @@ class JvmBuiltinOptimizationLowering(val context: JvmBackendContext) : FileLower
                     expression
                 }
             }
+
+            override fun visitWhen(expression: IrWhen): IrExpression {
+                expression.transformChildrenVoid(this)
+                // Remove all branches with constant false condition.
+                expression.branches.removeIf() {
+                    it.condition.isFalseConst()
+                }
+                // If the only condition that is left has a constant true condition remove the
+                // when in favor of the result. If there are no conditions left, remove the when
+                // entirely and replace it with an empty block.
+                return if (expression.branches.size == 0) {
+                    IrBlockImpl(expression.startOffset, expression.endOffset, context.irBuiltIns.unitType)
+                } else {
+                    expression.branches.first().takeIf { it.condition.isTrueConst() }?.result ?: expression
+                }
+            }
+
+            private fun isImmutableTemporaryVariableWithConstantValue(statement: IrStatement): Boolean {
+                return statement is IrVariable &&
+                        statement.origin == IrDeclarationOrigin.IR_TEMPORARY_VARIABLE &&
+                        !statement.isVar &&
+                        statement.initializer is IrConst<*>
+            }
+
+            override fun visitBlock(expression: IrBlock): IrExpression {
```

Comment:
If I understanding correctly visitBlock/visitGetValue transformations are required to clean redundant bytecode after other transformation in this lower? If true please mention it in code with some examples

Reply:
That is correct. I have added fairly large comments to explain what is going on. Let me know what you think.